rfcs/text/0004_application_service_mounting.md

```diff
@@ -0,0 +1,237 @@
+- Start Date: 2019-05-10
+- RFC PR: (leave this empty)
+- Kibana Issue: (leave this empty)
+
+# Summary
+
+A front-end service to manage registration and root-level routing for
+first-class applications.
+
+# Basic example
+
+```tsx
+import React from 'react';
+import ReactDOM from 'react-dom';
+
+import { MyApp } from './components';
+
+class MyPlugin {
+  setup({ application }) {
+    application.register({
+      id: 'my-app',
+      title: 'My Application',
+      mount(context, targetDomElement) {
+        ReactDOM.render(
+          <MyAppRoot mountContext={context} deps={pluginStart} />,
+          targetDomElement
+        );
+
+        return () => {
```

Comment:
@joshdover Based on our conversation about catching and rendering app-level errors, what happens to this callback in such a scenario? If the app fails and we render an error in place of the app, then I'm guessing we'll call this callback at that point? I can't think of any problems with this, just wanted to bring it to your attention.

Reply:
Yes, I would expect that this gets called immediately before the error UI is displayed.